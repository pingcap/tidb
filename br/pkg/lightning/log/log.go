// Copyright 2019 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package log

import (
	"context"
	"time"

	"github.com/pingcap/errors"
	pclog "github.com/pingcap/log"
	"github.com/pingcap/tidb/util/logutil"
	"go.uber.org/zap"
	"go.uber.org/zap/zapcore"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

const (
	defaultLogLevel   = "info"
	defaultLogMaxDays = 7
	defaultLogMaxSize = 512 // MB
)

// Config serializes log related config in toml/json.
type Config struct {
	// Log level.
	Level string `toml:"level" json:"level"`
	// Log filename, leave empty to disable file log.
	File string `toml:"file" json:"file"`
	// Max size for a single file, in MB.
	FileMaxSize int `toml:"max-size" json:"max-size"`
	// Max log keep days, default is never deleting.
	FileMaxDays int `toml:"max-days" json:"max-days"`
	// Maximum number of old log files to retain.
	FileMaxBackups int `toml:"max-backups" json:"max-backups"`
}

// Adjust adjusts some fields in the config to a proper value.
func (cfg *Config) Adjust() {
	if len(cfg.Level) == 0 {
		cfg.Level = defaultLogLevel
	}
	if cfg.Level == "warning" {
		cfg.Level = "warn"
	}
	if cfg.FileMaxSize == 0 {
		cfg.FileMaxSize = defaultLogMaxSize
	}
	if cfg.FileMaxDays == 0 {
		cfg.FileMaxDays = defaultLogMaxDays
	}
}

// Logger is a simple wrapper around *zap.Logger which provides some extra
// methods to simplify Lightning's log usage.
type Logger struct {
	*zap.Logger
}

// logger for lightning, different from tidb logger.
var (
	appLogger = Logger{zap.NewNop()}
	appLevel  = zap.NewAtomicLevel()
)

// InitLogger initializes Lightning's and also the TiDB library's loggers.
func InitLogger(cfg *Config, _ string) error {
	tidbLogCfg := logutil.LogConfig{}
	// Disable annoying TiDB Log.
	// TODO: some error logs outputs randomly, we need to fix them in TiDB.
	tidbLogCfg.Level = "fatal"
	err := logutil.InitLogger(&tidbLogCfg)
	if err != nil {
		return errors.Trace(err)
	}

	logCfg := &pclog.Config{
		Level:         cfg.Level,
		DisableCaller: false, // FilterCore requires zap.AddCaller.
	}
	filterTiDBLog := zap.WrapCore(func(core zapcore.Core) zapcore.Core {
		// Filter logs from TiDB and PD.
		return NewFilterCore(core, "github.com/pingcap/tidb/br/", "main.main")
	})
	// "-" is a special config for log to stdout.
	if len(cfg.File) > 0 && cfg.File != "-" {
		logCfg.File = pclog.FileLogConfig{
			Filename:   cfg.File,
			MaxSize:    cfg.FileMaxSize,
			MaxDays:    cfg.FileMaxDays,
			MaxBackups: cfg.FileMaxBackups,
		}
	}
	logger, props, err := pclog.InitLogger(logCfg, filterTiDBLog)
	if err != nil {
		return err
	}
	pclog.ReplaceGlobals(logger, props)

	// Do not log stack traces at all, as we'll get the stack trace from the
	// error itself.
	appLogger = Logger{logger.WithOptions(zap.AddStacktrace(zap.DPanicLevel))}
	appLevel = props.Level

	return nil
}

// SetAppLogger replaces the default logger in this package to given one
func SetAppLogger(l *zap.Logger) {
	appLogger = Logger{l.WithOptions(zap.AddStacktrace(zap.DPanicLevel))}
}

// L returns the current logger for Lightning.
func L() Logger {
	return appLogger
}

// Level returns the current global log level.
func Level() zapcore.Level {
	return appLevel.Level()
}

// SetLevel modifies the log level of the global logger. Returns the previous
// level.
func SetLevel(level zapcore.Level) zapcore.Level {
	oldLevel := appLevel.Level()
	appLevel.SetLevel(level)
	return oldLevel
}

// ShortError contructs a field which only records the error message without the
// verbose text (i.e. excludes the stack trace).
//
// In Lightning, all errors are almost always propagated back to `main()` where
// the error stack is written. Including the stack in the middle thus usually
// just repeats known information. You should almost always use `ShortError`
// instead of `zap.Error`, unless the error is no longer propagated upwards.
func ShortError(err error) zap.Field {
	if err == nil {
		return zap.Skip()
	}
	return zap.String("error", err.Error())
}

// With creates a child logger from the global logger and adds structured
// context to it.
func With(fields ...zap.Field) Logger {
	return appLogger.With(fields...)
}

// IsContextCanceledError returns whether the error is caused by context
// cancellation.
func IsContextCanceledError(err error) bool {
	err = errors.Cause(err)
	return err == context.Canceled || status.Code(err) == codes.Canceled
}

// Begin marks the beginning of a task.
func (logger Logger) Begin(level zapcore.Level, name string) *Task {
	if ce := logger.WithOptions(zap.AddCallerSkip(1)).Check(level, name+" start"); ce != nil {
		ce.Write()
	}
	return &Task{
		Logger: logger,
		level:  level,
		name:   name,
		since:  time.Now(),
	}
}

// With creates a child logger and adds structured context to it.
func (logger Logger) With(fields ...zap.Field) Logger {
	return Logger{logger.Logger.With(fields...)}
}

// Named adds a new path segment to the logger's name.
func (logger Logger) Named(name string) Logger {
	return Logger{logger.Logger.Named(name)}
}

// Task is a logger for a task spanning a period of time. This structure records
// when the task is started, so the time elapsed for the whole task can be
// logged without book-keeping.
type Task struct {
	Logger
	level zapcore.Level
	name  string
	since time.Time
}

// End marks the end of a task.
//
// The `level` is the log level if the task *failed* (i.e. `err != nil`). If the
// task *succeeded* (i.e. `err == nil`), the level from `Begin()` is used
// instead.
//
// The `extraFields` are included in the log only when the task succeeded.
func (task *Task) End(level zapcore.Level, err error, extraFields ...zap.Field) time.Duration {
	elapsed := time.Since(task.since)
	var verb string
	switch {
	case err == nil:
		level = task.level
		verb = " completed"
	case IsContextCanceledError(err):
		level = zap.DebugLevel
		verb = " canceled"
		extraFields = nil
	default:
		verb = " failed"
		extraFields = nil
	}
	if ce := task.WithOptions(zap.AddCallerSkip(1)).Check(level, task.name+verb); ce != nil {
		ce.Write(append(
			extraFields,
			zap.Duration("takeTime", elapsed),
			ShortError(err),
		)...)
	}
	return elapsed
}

type ctxKeyType struct{}

var ctxKey ctxKeyType

// NewContext returns a new context with the provided logger.
func NewContext(ctx context.Context, logger Logger) context.Context {
	return context.WithValue(ctx, ctxKey, logger)
}

// FromContext returns the logger stored in the context.
func FromContext(ctx context.Context) Logger {
	m, ok := ctx.Value(ctxKey).(Logger)
	if !ok {
		return appLogger
	}
	return m
}
