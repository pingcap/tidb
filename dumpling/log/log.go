// Copyright 2020 PingCAP, Inc. Licensed under Apache-2.0.

package log

import (
	"github.com/pingcap/errors"
	pclog "github.com/pingcap/log"
	"go.uber.org/zap"
)

var appLogger = Logger{zap.NewNop()}

// Logger wraps the zap logger.
type Logger struct {
	*zap.Logger
}

// Zap returns the global logger.
func Zap() Logger {
	return appLogger
}

// Config serializes log related config in toml/json.
type Config struct {
	// Log level.
	// One of "debug", "info", "warn", "error", "dpanic", "panic", and "fatal".
	Level string `toml:"level" json:"level"`
	// Log filename, leave empty to disable file log.
	File string `toml:"file" json:"file"`
	// Max size for a single file, in MB.
	FileMaxSize int `toml:"max-size" json:"max-size"`
	// Max log keep days, default is never deleting.
	FileMaxDays int `toml:"max-days" json:"max-days"`
	// Maximum number of old log files to retain.
	FileMaxBackups int `toml:"max-backups" json:"max-backups"`
	// Format of the log, one of `text`, `json` or `console`.
	Format string `toml:"format" json:"format"`
}

// InitAppLogger inits the wrapped logger from config.
func InitAppLogger(cfg *Config) (Logger, *pclog.ZapProperties, error) {
	logger, props, err := pclog.InitLogger(&pclog.Config{
		Level: cfg.Level,
		File: pclog.FileLogConfig{
			Filename:   cfg.File,
			MaxSize:    cfg.FileMaxSize,
			MaxDays:    cfg.FileMaxDays,
			MaxBackups: cfg.FileMaxBackups,
		},
		Format: cfg.Format,
	})
	if err != nil {
		return appLogger, props, errors.Trace(err)
	}
	return Logger{logger.WithOptions(zap.AddStacktrace(zap.DPanicLevel))}, props, nil
}

// NewAppLogger returns the wrapped logger from config.
func NewAppLogger(logger *zap.Logger) Logger {
	return Logger{logger}
}

// ShortError contructs a field which only records the error message without the
// verbose text (i.e. excludes the stack trace).
func ShortError(err error) zap.Field {
	if err == nil {
		return zap.Skip()
	}
	return zap.String("error", err.Error())
}
