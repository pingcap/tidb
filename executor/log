+ go test . --tags=intest -v -run TestIndexMergePanicPartialIndexWorker
