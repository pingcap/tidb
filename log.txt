cat checklist.md
# Following the checklist saves the reviewers' time and gets your PR reviewed faster.

# Self Review
Have you reviewed every line of your changes by yourself?

# Test
Have you added enough test cases to cover the new feature or bug fix?
Also, add comments to describe your test cases.

# Naming
Do function names keep consistent with its behavior?
Is it easy to infer the function's behavior by its name?

# Comment
Is there any code that confuses the reviewer?
Add comments on them! You'll be asked to do so anyway.
Make sure there is no syntax or spelling error in your comments.
Some online syntax checking tools like Grammarly may be helpful.

# Refactor
Is there any way to refactor the code to make it more readable?
If the refactoring touches a lot of existing code, send another PR to do it.

# Single Purpose
Make sure the PR does only one thing and nothing else.

# Diff Size
Make sure the diff size is no more than 500, split it into small PRs if it is too large.
go get github.com/coreos/gofail
Running in native mode.
ok  	github.com/pingcap/tidb/ast	(cached)	coverage: 76.9% of statements
?   	github.com/pingcap/tidb/cmd/benchdb	[no test files]
?   	github.com/pingcap/tidb/cmd/benchfilesort	[no test files]
?   	github.com/pingcap/tidb/cmd/benchkv	[no test files]
?   	github.com/pingcap/tidb/cmd/benchraw	[no test files]
?   	github.com/pingcap/tidb/cmd/explaintest	[no test files]
ok  	github.com/pingcap/tidb/cmd/importer	(cached)	coverage: 1.5% of statements
ok  	github.com/pingcap/tidb/config	(cached)	coverage: 14.3% of statements
ok  	github.com/pingcap/tidb/ddl	(cached)	coverage: 79.2% of statements
?   	github.com/pingcap/tidb/ddl/util	[no test files]
ok  	github.com/pingcap/tidb/distsql	(cached)	coverage: 69.7% of statements
ok  	github.com/pingcap/tidb/domain	(cached)	coverage: 47.9% of statements
ok  	github.com/pingcap/tidb/executor	(cached)	coverage: 84.2% of statements
ok  	github.com/pingcap/tidb/expression	(cached)	coverage: 70.0% of statements
ok  	github.com/pingcap/tidb/expression/aggregation	(cached)	coverage: 63.5% of statements
ok  	github.com/pingcap/tidb/infoschema	(cached)	coverage: 50.3% of statements
ok  	github.com/pingcap/tidb/kv	(cached)	coverage: 70.8% of statements
ok  	github.com/pingcap/tidb/meta	(cached)	coverage: 79.9% of statements
ok  	github.com/pingcap/tidb/meta/autoid	(cached)	coverage: 75.6% of statements
ok  	github.com/pingcap/tidb/metrics	(cached)	coverage: 90.7% of statements
ok  	github.com/pingcap/tidb/model	(cached)	coverage: 64.6% of statements
ok  	github.com/pingcap/tidb/mysql	(cached)	coverage: 48.2% of statements
ok  	github.com/pingcap/tidb/owner	(cached)	coverage: 24.0% of statements
ok  	github.com/pingcap/tidb/parser	(cached)	coverage: 91.1% of statements
?   	github.com/pingcap/tidb/parser/goyacc	[no test files]
ok  	github.com/pingcap/tidb/parser/opcode	(cached)	coverage: 80.0% of statements
ok  	github.com/pingcap/tidb/perfschema	(cached)	coverage: 87.2% of statements
ok  	github.com/pingcap/tidb/plan	(cached)	coverage: 80.4% of statements
?   	github.com/pingcap/tidb/privilege	[no test files]
ok  	github.com/pingcap/tidb/privilege/privileges	(cached)	coverage: 76.1% of statements
ok  	github.com/pingcap/tidb/server	(cached)	coverage: 62.4% of statements
ok  	github.com/pingcap/tidb/session	(cached)	coverage: 82.4% of statements
?   	github.com/pingcap/tidb/sessionctx	[no test files]
ok  	github.com/pingcap/tidb/sessionctx/binloginfo	(cached)	coverage: 83.3% of statements
?   	github.com/pingcap/tidb/sessionctx/stmtctx	[no test files]
ok  	github.com/pingcap/tidb/sessionctx/variable	(cached)	coverage: 50.8% of statements
ok  	github.com/pingcap/tidb/statistics	(cached)	coverage: 83.6% of statements
ok  	github.com/pingcap/tidb/store	(cached)	coverage: 0.0% of statements
?   	github.com/pingcap/tidb/store/mockoracle	[no test files]
?   	github.com/pingcap/tidb/store/mockstore	[no test files]
ok  	github.com/pingcap/tidb/store/mockstore/mocktikv	(cached)	coverage: 29.4% of statements
ok  	github.com/pingcap/tidb/store/tikv	18.057s	coverage: 72.4% of statements
ok  	github.com/pingcap/tidb/store/tikv/gcworker	(cached)	coverage: 52.6% of statements
ok  	github.com/pingcap/tidb/store/tikv/latch	(cached)	coverage: 89.8% of statements
?   	github.com/pingcap/tidb/store/tikv/oracle	[no test files]
ok  	github.com/pingcap/tidb/store/tikv/oracle/oracles	(cached)	coverage: 21.0% of statements
?   	github.com/pingcap/tidb/store/tikv/tikvrpc	[no test files]
ok  	github.com/pingcap/tidb/structure	(cached)	coverage: 82.9% of statements
ok  	github.com/pingcap/tidb/table	(cached)	coverage: 88.0% of statements
ok  	github.com/pingcap/tidb/table/tables	(cached)	coverage: 71.6% of statements
ok  	github.com/pingcap/tidb/tablecodec	(cached)	coverage: 58.3% of statements
ok  	github.com/pingcap/tidb/terror	(cached)	coverage: 79.7% of statements
?   	github.com/pingcap/tidb/tidb-server	[no test files]
ok  	github.com/pingcap/tidb/types	(cached)	coverage: 78.6% of statements
ok  	github.com/pingcap/tidb/types/json	(cached)	coverage: 81.1% of statements
ok  	github.com/pingcap/tidb/util	(cached)	coverage: 56.4% of statements
ok  	github.com/pingcap/tidb/util/admin	(cached)	coverage: 81.6% of statements
ok  	github.com/pingcap/tidb/util/arena	(cached)	coverage: 100.0% of statements
ok  	github.com/pingcap/tidb/util/auth	(cached)	coverage: 0.0% of statements [no tests to run]
ok  	github.com/pingcap/tidb/util/charset	(cached)	coverage: 61.5% of statements
ok  	github.com/pingcap/tidb/util/chunk	(cached)	coverage: 81.0% of statements
ok  	github.com/pingcap/tidb/util/codec	(cached)	coverage: 75.6% of statements
ok  	github.com/pingcap/tidb/util/encrypt	(cached)	coverage: 90.1% of statements
ok  	github.com/pingcap/tidb/util/filesort	(cached)	coverage: 84.0% of statements
ok  	github.com/pingcap/tidb/util/format	(cached)	coverage: 70.9% of statements
ok  	github.com/pingcap/tidb/util/goroutine_pool	(cached)	coverage: 100.0% of statements
ok  	github.com/pingcap/tidb/util/hack	(cached)	coverage: 92.3% of statements
?   	github.com/pingcap/tidb/util/israce	[no test files]
ok  	github.com/pingcap/tidb/util/kvcache	(cached)	coverage: 54.3% of statements
ok  	github.com/pingcap/tidb/util/kvencoder	(cached)	coverage: 75.5% of statements
ok  	github.com/pingcap/tidb/util/logutil	0.016s	coverage: 75.8% of statements
ok  	github.com/pingcap/tidb/util/memory	(cached)	coverage: 82.9% of statements
ok  	github.com/pingcap/tidb/util/mock	(cached)	coverage: 10.5% of statements
ok  	github.com/pingcap/tidb/util/mvmap	(cached)	coverage: 86.6% of statements
ok  	github.com/pingcap/tidb/util/printer	(cached)	coverage: 79.1% of statements
ok  	github.com/pingcap/tidb/util/ranger	(cached)	coverage: 79.6% of statements
?   	github.com/pingcap/tidb/util/sqlexec	[no test files]
ok  	github.com/pingcap/tidb/util/stringutil	(cached)	coverage: 100.0% of statements
ok  	github.com/pingcap/tidb/util/systimemon	(cached)	coverage: 100.0% of statements
ok  	github.com/pingcap/tidb/util/testkit	(cached)	coverage: 14.0% of statements
?   	github.com/pingcap/tidb/util/testleak	[no test files]
ok  	github.com/pingcap/tidb/util/testutil	(cached)	coverage: 46.5% of statements
?   	github.com/pingcap/tidb/x-server	[no test files]
CGO_ENABLED=0 go build   -ldflags '-X "github.com/pingcap/tidb/mysql.TiDBReleaseVersion=v2.1.0-alpha-7-g41ec1e88-dirty" -X "github.com/pingcap/tidb/util/printer.TiDBBuildTS=2018-06-05 09:31:27" -X "github.com/pingcap/tidb/util/printer.TiDBGitHash=41ec1e88986fd993f613341ec04fbb74dc217369" -X "github.com/pingcap/tidb/util/printer.TiDBGitBranch=dev/ifconstantfold" -X "github.com/pingcap/tidb/util/printer.GoVersion=go version go1.10.1 darwin/amd64"' -o bin/tidb-server tidb-server/main.go
skip building tidb-server, using existing binary: ../../bin/tidb-server
skip building importer, using existing binary: 
/Users/liuzhuomin/go:/Users/liuzhuomin/go/src/github.com/pingcap/tidb/vendor
building explain-test binary: ./explain_test
start tidb-server, log file: ./explain-test.out
tidb-server(PID: 13345) started
run all explain test cases
