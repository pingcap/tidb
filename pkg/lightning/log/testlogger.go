// Copyright 2019 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package log

import (
	"go.uber.org/zap"
	"go.uber.org/zap/zapcore"
	"go.uber.org/zap/zaptest"
)

// MakeTestLogger creates a Logger instance which produces JSON logs.
func MakeTestLogger(opts ...zap.Option) (Logger, *zaptest.Buffer) {
	buffer := new(zaptest.Buffer)
	logger := zap.New(zapcore.NewCore(
		zapcore.NewJSONEncoder(zapcore.EncoderConfig{
			LevelKey:       "$lvl",
			MessageKey:     "$msg",
			EncodeLevel:    zapcore.CapitalLevelEncoder,
			EncodeDuration: zapcore.StringDurationEncoder,
		}),
		buffer,
		zap.DebugLevel,
	), opts...)
	return Logger{Logger: logger}, buffer
}
