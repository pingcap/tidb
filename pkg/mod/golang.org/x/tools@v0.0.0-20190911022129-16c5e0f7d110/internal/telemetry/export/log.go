// Copyright 2019 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package export

import (
	"context"
	"fmt"
	"io"

	"golang.org/x/tools/internal/telemetry"
)

// LogWriter returns an observer that logs events to the supplied writer.
// If onlyErrors is true it does not log any event that did not have an
// associated error.
// It ignores all telemetry other than log events.
func LogWriter(w io.Writer, onlyErrors bool) Exporter {
	return &logWriter{writer: w, onlyErrors: onlyErrors}
}

type logWriter struct {
	writer     io.Writer
	onlyErrors bool
}

func (w *logWriter) StartSpan(context.Context, *telemetry.Span)  {}
func (w *logWriter) FinishSpan(context.Context, *telemetry.Span) {}
func (w *logWriter) Log(ctx context.Context, event telemetry.Event) {
	if event.Error == nil {
		// we only log errors by default
		return
	}
	fmt.Fprintf(w.writer, "%v\n", event)
}
func (w *logWriter) Metric(context.Context, telemetry.MetricData) {}
func (w *logWriter) Flush()                                       {}
