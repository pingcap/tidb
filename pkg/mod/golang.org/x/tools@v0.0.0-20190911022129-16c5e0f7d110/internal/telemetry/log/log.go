// Copyright 2019 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

// Package log is a context based logging package, designed to interact well
// with both the lsp protocol and the other telemetry packages.
package log

import (
	"context"
	"time"

	"golang.org/x/tools/internal/telemetry"
	"golang.org/x/tools/internal/telemetry/export"
	"golang.org/x/tools/internal/telemetry/tag"
)

type Event telemetry.Event

// With sends a tag list to the installed loggers.
func With(ctx context.Context, tags ...telemetry.Tag) {
	export.Log(ctx, telemetry.Event{
		At:   time.Now(),
		Tags: tags,
	})
}

// Print takes a message and a tag list and combines them into a single tag
// list before delivering them to the loggers.
func Print(ctx context.Context, message string, tags ...tag.Tagger) {
	export.Log(ctx, telemetry.Event{
		At:      time.Now(),
		Message: message,
		Tags:    tag.Tags(ctx, tags...),
	})
}

// Print takes a message and a tag list and combines them into a single tag
// list before delivering them to the loggers.
func Error(ctx context.Context, message string, err error, tags ...tag.Tagger) {
	if err == nil {
		err = errorString(message)
		message = ""
	}
	export.Log(ctx, telemetry.Event{
		At:      time.Now(),
		Message: message,
		Error:   err,
		Tags:    tag.Tags(ctx, tags...),
	})
}

type errorString string

// Error allows errorString to conform to the error interface.
func (err errorString) Error() string { return string(err) }
