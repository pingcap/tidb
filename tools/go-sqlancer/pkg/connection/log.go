// Copyright 2022 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package connection

import (
	"fmt"
	"time"

	"github.com/pingcap/log"
	"go.uber.org/zap"
)

func (c *Connection) logSQL(sql string, duration time.Duration, err error, args ...interface{}) {
	line := fmt.Sprintf("Success: %t, Duration: %s", err == nil, duration)
	for index, arg := range args {
		if index == 0 {
			if affectedRows, ok := arg.(int64); ok {
				line = fmt.Sprintf("%s, Affected Rows: %d", line, affectedRows)
			}
		}
	}
	if err := c.logger.Infof("%s, SQL: %s", line, sql); err != nil {
		log.Fatal("fail to log to file %v", zap.Error(err))
	}
}
