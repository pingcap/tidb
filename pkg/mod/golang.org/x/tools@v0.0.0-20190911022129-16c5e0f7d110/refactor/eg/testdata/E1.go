// +build ignore

package E1

import "log"

func example() {
	log.Fatal("oops") // match
}
