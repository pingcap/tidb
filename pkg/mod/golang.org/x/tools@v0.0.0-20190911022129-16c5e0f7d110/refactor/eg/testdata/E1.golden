// +build ignore

package E1

import (
	"fmt"
	"log"
	"os"
)

func example() {
	fmt.Fprintf(os.Stderr, "warning: %v", "oops") // match
}
